### ISSUE TITLE
Rollback optimistic txn if meeting an unexpected primary lock when resolving locks for write

### ISSUE THREAD

**Issue by sticnarf** [tikv/tikv#11148]
## Problem

Although we have pessimistic transactions now, optimistic transactions tend to be more efficient if there are few conflicts. And it's also why TiDB prefers optimistic transactions for auto-commit transactions.

When there are write conflicts among optimistic transactions, one of the conflicted transactions must be rolled back. The rollback usually happens fast, which is expected. But sometimes multiple optimistic transactions will form deadlocks, then rollback will happen only when one lock exceeds its TTL. The TTL is usually 3 seconds, causing all the involving transactions to be stuck for 3 seconds.

Here is a concrete example:

Both txn1 (start_ts = 10) and txn2 (start_ts = 20) writes `k1` and `k2`. And the primary lock of these two transactions is `k1`. 

1. Txn1 locks `k1`.
2. Txn2 locks `k2`.
3. Txn1 wants to lock `k2` but find a lock from txn2. Txn1 queries the primary lock of txn2 but it does not exist (`k1` is locked by txn1 itself). Now it keeps CheckTxnStatus `k1` until found the lock of txn2. https://github.com/tikv/client-go/blob/a5250aa7c446e06f0a05d761c5a2c3923f8b48e2/txnkv/txnlock/lock_resolver.go#L489-L492
4. Txn2 wants to lock `k1` but find a lock from txn1. Txn2 queries the primary lock of txn1 but finds it is not expired. So it keeps retry until it expires.

Note that the transaction who has locked the primary lock of another transaction must have a smaller start_ts. Otherwise, the transaction with smaller start_ts will meet a lock with larger ts when it is writing its primary lock, it will abort itself.

## Enhancement

If the primary lock is not expected, we should rollback the lock of one of the transactions when we are resolving for write.

The simpliest idea is just to rollback the lock we want to resolve. But this will rollback the transaction with larger start ts. Still using the above example, `k2` will be rolled back and there will be a rollback record with commit_ts = 20. Then, both of the transactions will be aborted. This is apparently not optimal.

So, I think a better solution is to rollback self if the primary lock is locked by the transaction itself.

As stated above, such a deadlock happens if the lock ts of the primary lock is smaller, so this rollback will only affect one transaction. It is more likely to be safe because we are rolling back self, so we don't need to worry that we will potentially break a committed async-commit transaction.

~~We can return the lock TS in `CheckTxnStatus`. Then, if the lock resolver finds the mismatching primary lock is the writer itself, the writer can abort itself.~~

Now, the solution is that we can directly abort the current transaction when an optimistic transaction encounters a lock with larger start TS.

**Comment by sticnarf** [comment]
- [ ] Add `lock_ts` field to `TxnNotFound` https://github.com/pingcap/kvproto/pull/832
- [ ] Return mismatching lock TS in `TxnNotFound` error in unistore https://github.com/pingcap/tidb/pull/29394
- [ ] Return mismatching lock TS in `TxnNotFound` error in TiKV https://github.com/tikv/tikv/pull/11208
- [ ] Abort if the mismatching lock is from the current transaction when resolving locks https://github.com/tikv/client-go/pull/367

**Comment by sticnarf** [comment]
@youjiali1995 @MyonKeminta PTAL the above PRs with order.

**Comment by MyonKeminta** [comment]
Are these PRs still drafts? The solution looks fine to me

**Comment by sticnarf** [comment]
> Are these PRs still drafts? The solution looks fine to me

No, they are reviewable. They are drafts only if there are patches to be replace.

**Comment by youjiali1995** [comment]
An optimistic transaction that meets a lock with larger ts must be rolled back due to write conflict. The reason why txn1 blocks here is it can't find the primary lock of txn2. Can we just make txn1 rolls back itself even if the primary lock of txn2 doesn't exsit?

**Comment by youjiali1995** [comment]
> An optimistic transaction that meets a lock with larger ts must be rolled back due to write conflict. The reason why txn1 blocks here is it can't find the primary lock of txn2. Can we just make txn1 rolls back itself even if the primary lock of txn2 doesn't exsit?

We only need to change the impelmentation of lock resolver and I think its effect is same as your solution.

**Comment by sticnarf** [comment]
> An optimistic transaction that meets a lock with larger ts must be rolled back due to write conflict. The reason why txn1 blocks here is it can't find the primary lock of txn2. Can we just make txn1 rolls back itself even if the primary lock of txn2 doesn't exsit?

I thought there might be slight difference in success rate. But maybe not so important...

If start ts of txn1 is 20 and start ts of txn2 is 10. Both of them write k1 and k2.

Txn2 first prewrites k2 but has not prewritten k1. Then, txn1 prewrites k2 and finds the lock. If the primary lock has not been prewritten yet, we don't need to abort the txn2 immediately. Then, both txn1 and txn2 can succeed.

**Comment by youjiali1995** [comment]
> If start ts of txn1 is 20 and start ts of txn2 is 10. Both of them write k1 and k2.
>
> Txn2 first prewrites k2 but has not prewritten k1. Then, txn1 prewrites k2 and finds the lock. If the primary lock has not been prewritten yet, we don't need to abort the txn2 immediately. Then, both txn1 and txn2 can succeed.

How can they are both committed? If txn2 doesn't meet locks of txn1, it can be committed and txn2 will be rolled back due to write conflict.

> And if the primary lock is from txn3 which will not cause deadlocks with txn1 and txn2, it can also be an unnecessary abort for txn1.

My point is a transaction meets a lock with larger ts can roll back itself directly, not to resolve other transactions. I think txn3 won't be affected.

**Comment by sticnarf** [comment]
> How can they are both committed? If txn2 doesn't meet locks of txn1, it can be committed and txn2 will be rolled back due to write conflict.

If txn2 then prewrites k1 with min_commit_ts=15, then txn2 can commit with commit_ts=15 using async commit. And txn1 can commit without write conflict.

**Comment by sticnarf** [comment]
Although check txn status will advance max ts, if the primary lock is from txn 3 (maybe start ts = 5), txn1 might not certainly abort.

**Comment by youjiali1995** [comment]
I guess I didn't make it clear and you misunstood it... 

The transaction with a small start_ts must be rolled back due to write conflict if it meets a lock with a larger ts. Is it right? Your solution is the transaction can roll back itself if it meets a lock with larger ts and the lock's primary lock is held by the transaction. The difference in my solution is the transction rolls back itself as long as it meets a lock with larger ts, no matter whether the lock's primary lock is held by the transction or not. 

Take your example for example,
> If start ts of txn1 is 20 and start ts of txn2 is 10. Both of them write k1 and k2.
> 
> Txn2 first prewrites k2 but has not prewritten k1. Then, txn1 prewrites k2 and finds the lock. If the primary lock has not been prewritten yet, we don't need to abort the txn2 immediately. Then, both txn1 and txn2 can succeed.
> 
> And if the primary lock is from txn3 which will not cause deadlocks with txn1 and txn2, it can also be an unnecessary abort for txn1.

My solution:
1. txn2(ts=10) locks k2.
2. txn1(ts=20) meets txn2's lock on k2 and waits.
3. txn3(ts=5) locks k1.
4. txn1(ts=20) meets txn3's lock on k1 and waits.
5. txn3 commits.
6. txn2 can commit.
7. txn1 can commit.

Take the example in description for example,
> Both txn1 (start_ts = 10) and txn2 (start_ts = 20) writes `k1` and `k2`. And the primary lock of these two transactions is `k1`. 
>
> 1. Txn1 locks `k1`.
> 2. Txn2 locks `k2`.
> 3. Txn1 wants to lock `k2` but find a lock from txn2. Txn1 queries the primary lock of txn2 but it does not exist (`k1` is locked by txn1 itself). Now it keeps CheckTxnStatus `k1` until found the lock of txn2. https://github.com/tikv/client-go/blob/a5250aa7c446e06f0a05d761c5a2c3923f8b48e2/txnkv/txnlock/lock_resolver.go#L489-L492
> 4. Txn2 wants to lock `k1` but find a lock from txn1. Txn2 queries the primary lock of txn1 but finds it is not expired. So it keeps retry until it expires.
> 
> Note that the transaction who has locked the primary lock of another transaction must have a smaller start_ts. Otherwise, the transaction with smaller start_ts will meet a lock with larger ts when it is writing its primary lock, it will abort itself.

My solution:
1. txn1(ts=10) locks k1.
2. txn2(ts=20) locks k2.
3. txn1(ts=10) meets txn2's lock on k2 and txn1 rolls back itself because its start_ts(10) < lock's ts(20).
4. txn2(ts=20) locks k1 and commit.

**Comment by sticnarf** [comment]
> Can we just make txn1 rolls back itself even if the primary lock of txn2 doesn't exsit?

I see... You mean this only when start ts of txn1 < txn2.

Well, then I think it does not even need the lock resolver. When prewrite encounters a lock with larger ts, it can abort itself immediately.

**Comment by youjiali1995** [comment]
> > Can we just make txn1 rolls back itself even if the primary lock of txn2 doesn't exsit?
> 
> I see... You mean this only when start ts of txn1 < txn2.
> 
> Well, then I think it does not even need the lock resolver. When prewrite encounters a lock with larger ts, it can abort itself immediately.

My fault. I didn't make it clear.😂 This sentence is in the context of the example in description, so txn1's start_ts < txn2's.
